tests/interlocked_win32_ut/interlocked_win32_ut.c

```diff
@@ -0,0 +1,708 @@
+    ASSERT_ARE_EQUAL(void_ptr, &value1, return_val);
+}
+
+/*Tests_SRS_INTERLOCKED_WIN32_43_021: [interlocked_decrement shall call InterlockedDecrement from windows.h.]*/
+/*Tests_SRS_INTERLOCKED_WIN32_43_022 : [interlocked_decrement shall return the resulting 32 - bit integer value.]*/
+TEST_FUNCTION(interlocked_decrement_calls_InterlockedDecrement)
+{
+    ///arrange
+    volatile int32_t addend = INT32_MAX;
+    STRICT_EXPECTED_CALL(mock_InterlockedDecrement((volatile LONG*)&addend))
+        .SetReturn(INT32_MAX -1);
+    ///act
+    int32_t return_val = interlocked_decrement(&addend);
+
+    ///assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    ASSERT_ARE_EQUAL(int32_t, INT32_MAX -1, return_val);
+}
+
+/*Tests_SRS_INTERLOCKED_WIN32_43_023: [interlocked_decrement_16 shall call InterlockedDecrement16 from windows.h.]*/
+/*Tests_SRS_INTERLOCKED_WIN32_43_024 : [interlocked_decrement_16 shall return the resulting 16 - bit integer value.]*/
+TEST_FUNCTION(interlocked_decrement_16_calls_InterlockedDecrement16)
+{
+    ///arrange
+    volatile int16_t addend = INT16_MAX;
+    STRICT_EXPECTED_CALL(mock_InterlockedDecrement16((volatile SHORT*)&addend))
+        .SetReturn(INT16_MAX -1);
+    ///act
+    int16_t return_val = interlocked_decrement_16(&addend);
+
+    ///assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    ASSERT_ARE_EQUAL(int16_t, INT16_MAX -1, return_val);
+}
+
+/*Tests_SRS_INTERLOCKED_WIN32_43_025: [interlocked_decrement_64 shall call InterlockedDecrement64 from windows.h.]*/
+/*Tests_SRS_INTERLOCKED_WIN32_43_026 : [interlocked_decrement_64 shall return the resulting 64 - bit integer value.]*/
+TEST_FUNCTION(interlocked_decrement_64_calls_InterlockedDecrement64)
+{
+    ///arrange
+    volatile int64_t addend = INT64_MAX;
+    STRICT_EXPECTED_CALL(mock_InterlockedDecrement64((volatile LONG64*)&addend))
+        .SetReturn(INT64_MAX - 1);
+    ///act
+    int64_t return_val = interlocked_decrement_64(&addend);
+
+    ///assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    ASSERT_ARE_EQUAL(int64_t, INT64_MAX - 1, return_val);
+}
+
+/*Tests_SRS_INTERLOCKED_WIN32_43_027: [interlocked_exchange shall call InterlockedExchange from windows.h.]*/
+/*Tests_SRS_INTERLOCKED_WIN32_43_028 : [interlocked_exchange shall return the initial value pointed to by target.]*/
+TEST_FUNCTION(interlocked_exchange_calls_InterlockedExchange)
+{
+    ///arrange
+    volatile int32_t target = INT32_MIN;
+    int32_t value = INT32_MAX;
+    STRICT_EXPECTED_CALL(mock_InterlockedExchange((volatile LONG*)&target, (LONG)value))
+        .SetReturn(INT32_MIN);
+
+    ///act
+    int32_t return_val = interlocked_exchange(&target, value);
+
+    ///assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
```

Comment:
>ASSERT_ARE_EQUAL [](start = 4, length = 16)

should assert target to be something.

Reply:
sort of a here and everywhere matter

---
In reply to: [427744935](https://github.com/Azure/azure-c-util/pull/43#discussion_r427744935) [](ancestors = 427744935)

Reply:
I'm doing that in the integration tests, so I didn't add that here. The INTERLOCKED_WIN32 requirements don't say anything about  target.

---
In reply to: [427745007](https://github.com/Azure/azure-c-util/pull/43#discussion_r427745007) [](ancestors = 427745007,427744935)